### PR TITLE
[WIP] [10.5.1] - Marker Clustering

### PR DESCRIPTION
**Pull request by corrigancd**
@szydan, @jccq, @johnnylopes , I have a prototype and would like some feedback. Below is a GIF with two Point type layers. One is a POI overlay based on our companies index. The other is a point type stored layer containing 1.7 million trees from an area in the US.

Point 1 - I've added a mouseover/mouseout to show the geohash rectangle that the cluster is based on, it's color is the one that is configured. The 268 cluster is from the companies search, while the 6370 cluster is from the trees stored layer. They currently use the same colour sheme, my thinking is that we'll need to in some way distinguish these as separate layers without the need to hover on each cluster? If so, I could look into some color ramp libraries that will generate something suitable on the fly based on the configured color (suggestions on how to do this would be very welcome).

Point 2 - The cluster is positioned at the **centre of gravity** of the geohash. The 6370 cluster happens to contain only data that is in the bottom left of the geohash. I think it makes sense to use the centre of gravity over the **centroid** of the geohash as it represents more clearly the location of most points. A drawback is that there could be some outliers (e.g. a small number of points in the top right of the geohash) that are masked. This seems to be a good reason to keep the mouseover/mouseout in place, or at least have it as an option.

Point 3 - There may be overlapping points in the geohashes where it has been chosen to show individual documents for (i.e. the points that are under the 2433 cluster at the  bottom). @jccq, indicated that these should undergo a secondary clustering, where the search icon could be used with a number beside points that have been secondarily clustered. I'll need some thinking on how to do this, e.g. algorithms, time for R&D, in order to implement it correctly, something is coming to mind is adapting the function in the leaflet marker clustering plugin for this type of purpose. Perhaps this one is a clustering phase 2 task. 

TwoClusteredPointLayersLookingTheSame

### CODE REVIEW

**Comment by jccq** [comment]
1) Hi Edwin, for aggregation of points we really should use icons, not colors. So just put A a with number (e.g. 232) with a small entity icon next to it. If no Icon available then .. a ball with a color automatically selected using the standard algorithm for color selection (which then has to be shown also in the legend) . We will assume coordinates are not all the same for different entities so what will happen most of the time if you have 2 types is that you'll see 2 balls .. possibly overlapping. You can then turn on and off a layer to get more clarify if you need it or zoom in.

silly mockup (icons next to the numbers in the balls)
image

again = ball icon is the same as the POI icon if it exists
ball color = same as icon color if it exists or assigned using kibana algorithm if it doesnt exist. In thsi case the layer should get an auto assigned square icon in the legend with the the same color

2) in theory its nice to see the boundary. But it's a trap! you this should not matter if you use a sensibly sized geohas. I think here in the example the geohash is *way* too wide  and points at a problem with the undelying scaling (in your example the entire map is split in maybe 6 huge areas!!!) . If treated propely should always be small enough not to matter really that they could be spread. Makes sense?  SO good news here there might be nothing to do.

Now .. this is a simplification because geohases allow you to express also points for which you "dont have certainty" e.g. you couldsave a low res geohash .. in this case it would be appropriate to show a map with big rectangle.. but we're willingly ignoring this for now i'd say? :)

3) we live with it.. you can turn off layets :)

so good nes.. just add icons to balls and colors coherent with ocons and we might be good to go? happy to talk!!!!

**Comment by corrigancd** [comment]
Trade Off on some zoom levels. Both taken at zoom level 3: 
below Precision = 2:
image
below Precision = 1: 
image

**Comment by corrigancd** [comment]
Tests written, working on a small issue before it can be merged.
image

**Comment by corrigancd** [comment]
Ok, this is it for the minute. Unit tests written, manually tested and refined. There is a phase 2 issue for documents with the same coordinates - https://sirensolutions.atlassian.net/browse/INVE-11633:

@szydan, it's ready for code review

markerClusteringWorking

File: public/utils.js [hunk 1]
```diff
@@ -223,6 +223,101 @@ define(function (require) {
...
+    processAggRespForMarkerClustering: function (aggChartData, geoFilter, limit, geoField) {
+      const docFilters = {
+        bool: {
+          should: []
+        }
+      };
+
+      let aggFeatures;
+      let totalNumberOfDocsToRetrieve = 0;
+      if (_.has(aggChartData, 'geoJson.features')) {
```
Comment (szydan) [inline public/utils.js:302]: I think *_.has* will return true even if the *geoJson.features* is lets say null
Might be better to use *_.get*

File: public/vislib/geoFilterHelper.js [hunk 2]
```diff
@@ -53,5 +53,33 @@ module.exports = {
...
+        shape: {
+          type: 'envelope',
+          coordinates: [
+            [topLeft.lon, topLeft.lat],
+            [bottomRight.lon, bottomRight.lat]
+          ]
+        }
+      };
+    } else {
+      console.warn('unexpected geotype: ' + geotype);
```
Comment (szydan) [inline public/vislib/geoFilterHelper.js:77]: not sure but maybe better to show the warning so the user is aware that there is a problem

File: public/vislib/icons/searchIcon.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ export const searchIcon = function (faIcon, color, size) {
   }
 
   return L.divIcon({
-    className: `search-icon`,
+    className: ``,
```
Comment (szydan) [inline public/vislib/icons/searchIcon.js:30]: is this needed ?

**Comment by jccq** [comment]
hay any chance to have  a little spinner while the aggreagtions load e.g.
in the icon on top right?


Giovanni Tummarello
CPO, Founder, Siren
A Unit 3, GTC, Mervue Business Park, Galway, H91 CR20, Ireland
D +353 91704885 <+353+91704885>
P +353 (0)91 704 885  <+353+(0)91+704+885>
M +353 87 1272938  <+353+87+1272938>
E giovanni.tummarello@siren.io  <giovanni.tummarello@siren.io>
W www.siren.io
<http://www.siren.io?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://twitter.com/sirensearch?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://www.youtube.com/channel/UCKGsC-vD28r7hW6T9QspKPA?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://vimeo.com/sirenio?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://www.facebook.com/sirensearch?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<https://www.linkedin.com/company/11117365?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>
<http://twitter.com/jccq?utm_source=WiseStamp&utm_medium=email&utm_term=&utm_content=&utm_campaign=signature>

**Comment by szydan** [comment]
After trying it live we discovered with @corrigancd that there are cases where requests are fired where they should NOT be 

* zooming in out where no single layer is selected 
* zooming in out where only some layers are selected would still send a query for one of the unselected layers 

We have to open a separate ticket and eliminate at least some of these extra requests before we can ship this one to PAX, in their env any extra not needed requests is a problem

**Comment by corrigancd** [comment]
Just made issues for those points guys:
@jccq - https://sirensolutions.atlassian.net/browse/INVE-11641
@szydan - https://sirensolutions.atlassian.net/browse/INVE-11642

File: public/vislib/icons/searchIcon.js (hunk 3 above)
Comment (szydan) [reply public/vislib/icons/searchIcon.js:30]: please make it a normal string not a template 
```
`` -> ''
```
and put a note why this is needed so nobody will be tempted to remove it 
and

**Comment by szydan** [comment]
all good not approving yet until we decide what could be done about extra requests

**Pull request merged**